cpp/src/arrow/array/array_nested.cc

```diff
@@ -351,122 +481,21 @@ Result<std::shared_ptr<Array>> FlattenListViewArray(const ListViewArrayT& list_v
-  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayBuilder> value_builder,
-                        MakeBuilder(value_type, pool));
-  RETURN_NOT_OK(value_builder->Reserve(sum_of_list_view_sizes));
-  auto list_builder = std::make_shared<ListBuilderType>(pool, value_builder, list_type);
-  RETURN_NOT_OK(list_builder->Reserve(list_view_data->length));
-
-  ArraySpan values{*list_view_data->child_data[0]};
-  const auto* in_validity_bitmap = list_view_data->GetValues<uint8_t>(0);
-  const auto* in_offsets = list_view_data->GetValues<offset_type>(1);
-  const auto* in_sizes = list_view_data->GetValues<offset_type>(2);
-  for (int64_t i = 0; i < list_view_data->length; ++i) {
-    const bool is_valid =
-        !in_validity_bitmap ||
-        bit_util::GetBit(in_validity_bitmap, list_view_data->offset + i);
-    const int64_t size = is_valid ? in_sizes[i] : 0;
-    RETURN_NOT_OK(list_builder->Append(is_valid, size));
-    RETURN_NOT_OK(value_builder->AppendArraySlice(values, in_offsets[i], size));
-  }
-  std::shared_ptr<ArrayData> list_array_data;
-  RETURN_NOT_OK(list_builder->FinishInternal(&list_array_data));
-  return list_array_data;
-}
-
-}  // namespace
-
-namespace internal {
-
-template <typename TYPE>
-inline void SetListData(VarLengthListLikeArray<TYPE>* self,
-                        const std::shared_ptr<ArrayData>& data,
-                        Type::type expected_type_id) {
-  ARROW_CHECK_EQ(data->buffers.size(), is_list_view(TYPE::type_id) ? 3 : 2);
-  ARROW_CHECK_EQ(data->type->id(), expected_type_id);
-  ARROW_CHECK_EQ(data->child_data.size(), 1);
-
-  self->Array::SetData(data);
-
-  self->list_type_ = checked_cast<const TYPE*>(data->type.get());
-  self->raw_value_offsets_ =
-      data->GetValuesSafe<typename TYPE::offset_type>(1, /*offset=*/0);
-  // BaseListViewArray::SetData takes care of setting raw_value_sizes_.
-
-  ARROW_CHECK_EQ(self->list_type_->value_type()->id(), data->child_data[0]->type->id());
-  DCHECK(self->list_type_->value_type()->Equals(data->child_data[0]->type));
-  self->values_ = MakeArray(self->data_->child_data[0]);
+void InstantiateFlattenListArray() {
+  (void)FlattenListArray<ListArray, false>;
+  (void)FlattenListArray<ListArray, true>;
+  (void)FlattenListArray<LargeListArray, false>;
+  (void)FlattenListArray<LargeListArray, true>;
+  (void)FlattenListArray<FixedSizeListArray, false>;
+  (void)FlattenListArray<FixedSizeListArray, true>;
+  (void)FlattenListViewArray<ListViewArray, false, false>;
+  (void)FlattenListViewArray<ListViewArray, false, true>;
+  (void)FlattenListViewArray<ListViewArray, true, false>;
+  (void)FlattenListViewArray<ListViewArray, true, true>;
+  (void)FlattenListViewArray<LargeListViewArray, false, false>;
+  (void)FlattenListViewArray<LargeListViewArray, false, true>;
+  (void)FlattenListViewArray<LargeListViewArray, true, false>;
+  (void)FlattenListViewArray<LargeListViewArray, true, true>;
```

Comment:
Some of these instantiations are required since only PyArrow would templatize the function with the last parameter (include_nulls) = true.